File: lib/graph.ts
import { adjacencyGraph, connectedCountryCodes } from "@/data/adjacency";
import { countryByCode } from "@/data/countries";
import type { Continent } from "@/lib/game-engine/types";

/**
 * Find the shortest path between two countries using BFS.
 * Returns the full path including start and end, or null if unreachable.
 */
export function findShortestPath(
  start: string,
  end: string,
  excludeCountries?: Set<string>
): string[] | null {
  if (start === end) return [start];
  if (!adjacencyGraph[start] || !adjacencyGraph[end]) return null;

  const queue: string[][] = [[start]];
  const visited = new Set<string>([start]);

  while (queue.length > 0) {
    const path = queue.shift()!;
    const current = path[path.length - 1];

    const neighbors = adjacencyGraph[current] || [];
    for (const neighbor of neighbors) {
      if (visited.has(neighbor)) continue;
      if (excludeCountries?.has(neighbor)) continue;

      const newPath = [...path, neighbor];

      if (neighbor === end) return newPath;

      visited.add(neighbor);
      queue.push(newPath);
    }
  }

  return null; // no path found
}

/**
 * Get the shortest path length (number of edges) between two countries.
 * Returns -1 if unreachable.
 */
export function getShortestDistance(start: string, end: string): number {
  const path = findShortestPath(start, end);
  return path ? path.length - 1 : -1;
}

/**
 * Check if two countries are direct neighbors.
 */
export function isValidNeighbor(current: string, candidate: string): boolean {
  const neighbors = adjacencyGraph[current];
  return neighbors ? neighbors.includes(candidate) : false;
}

/**
 * Get all direct neighbors of a country.
 */
export function getNeighbors(countryCode: string): string[] {
  return adjacencyGraph[countryCode] || [];
}

/**
 * Check if a country has land borders (can participate in Connect mode).
 */
export function hasLandBorders(countryCode: string): boolean {
  return connectedCountryCodes.has(countryCode);
}

/**
 * Check if a country belongs to the given continent filter.
 */
function matchesContinent(code: string, continent: Continent): boolean {
  if (continent === "all") return true;
  const country = countryByCode[code];
  return country?.continent === continent;
}

/**
 * Find all countries at exactly a given BFS distance from start.
 * If continent is specified, only traverses countries on that continent.
 */
export function countriesAtDistance(
  start: string,
  distance: number,
  continent: Continent = "all"
): string[] {
  if (!adjacencyGraph[start]) return [];

  const visited = new Set<string>([start]);
  let frontier = [start];
  let currentDistance = 0;

  while (currentDistance < distance && frontier.length > 0) {
    const nextFrontier: string[] = [];
    for (const country of frontier) {
      const neighbors = adjacencyGraph[country] || [];
      for (const neighbor of neighbors) {
        if (!visited.has(neighbor) && matchesContinent(neighbor, continent)) {
          visited.add(neighbor);
          nextFrontier.push(neighbor);
        }
      }
    }
    frontier = nextFrontier;
    currentDistance++;
  }

  return frontier;
}

/**
 * Get a random connected country code (has land borders).
 * If continent is specified, only picks from that continent.
 */
export function getRandomConnectedCountry(continent: Continent = "all"): string {
  const codes = Array.from(connectedCountryCodes).filter((code) =>
    matchesContinent(code, continent)
  );
  if (codes.length === 0) {
    // Fallback to any connected country
    const all = Array.from(connectedCountryCodes);
    return all[Math.floor(Math.random() * all.length)];
  }
  return codes[Math.floor(Math.random() * codes.length)];
}

File: geoplay-mobile/lib/core/data/graph.dart
import 'dart:collection';
import 'adjacency.dart';

/// BFS pathfinding between two countries using the adjacency graph.
/// Ported from lib/graph.ts.

/// Find the shortest path between two countries using BFS.
/// Returns the path as a list of ISO codes (including start and end),
/// or null if no path exists.
List<String>? findShortestPath(String startCode, String endCode, {Set<String>? allowedCountries}) {
  if (startCode == endCode) return [startCode];
  if (!hasLandBorders(startCode) || !hasLandBorders(endCode)) return null;

  final visited = <String>{};
  final queue = Queue<List<String>>();

  visited.add(startCode);
  queue.add([startCode]);

  while (queue.isNotEmpty) {
    final path = queue.removeFirst();
    final current = path.last;

    for (final neighbor in getNeighbors(current)) {
      if (allowedCountries != null && !allowedCountries.contains(neighbor) && neighbor != endCode) {
         continue; 
      }
      if (neighbor == endCode) {
        return [...path, neighbor];
      }
      if (!visited.contains(neighbor)) {
        visited.add(neighbor);
        queue.add([...path, neighbor]);
      }
    }
  }

  return null; // No path found
}

/// Find all shortest paths between two countries (may be multiple).
List<List<String>> findAllShortestPaths(String startCode, String endCode) {
  if (startCode == endCode) return [[startCode]];
  if (!hasLandBorders(startCode) || !hasLandBorders(endCode)) return [];

  final results = <List<String>>[];
  final visited = <String>{};
  final queue = Queue<List<String>>();
  int? shortestLength;

  visited.add(startCode);
  queue.add([startCode]);

  while (queue.isNotEmpty) {
    final path = queue.removeFirst();

    // If we've found paths and this one is longer, stop.
    if (shortestLength != null && path.length > shortestLength) break;

    final current = path.last;

    for (final neighbor in getNeighbors(current)) {
      if (neighbor == endCode) {
        final completePath = [...path, neighbor];
        shortestLength ??= completePath.length;
        if (completePath.length == shortestLength) {
          results.add(completePath);
        }
      } else if (!visited.contains(neighbor)) {
        visited.add(neighbor);
        queue.add([...path, neighbor]);
      }
    }
  }

  return results;
}

/// Check if a path between two countries exists.
bool pathExists(String startCode, String endCode) {
  return findShortestPath(startCode, endCode) != null;
}
